lib/timers.js

```diff
@@ -477,47 +477,109 @@ exports.clearImmediate = function(immediate) {
 
 var unrefList, unrefTimer;
 
+function _makeTimerTimeout(timer) {
+  var domain = timer.domain;
+  var msecs = timer._idleTimeout;
+
+  // Timer has been unenrolled by another timer that fired at the same time,
+  // so don't make it timeout.
+  if (!msecs || msecs < 0)
+    return;
+
+  if (!timer._onTimeout)
+    return;
+
+  if (domain) {
+    if (domain._disposed)
+      return;
+
+    domain.enter();
+  }
+
+  debug('unreftimer firing timeout');
+  L.remove(timer);
+  timer._called = true;
+  _runOnTimeout(timer);
+
+  if (domain)
+    domain.exit();
+}
+
+function _runOnTimeout(timer) {
+  try {
+    var threw = true;
+    timer._onTimeout();
+    threw = false;
+  } finally {
+    if (threw) process.nextTick(unrefTimeout);
+  }
+}
 
 function unrefTimeout() {
   var now = Timer.now();
 
   debug('unrefTimer fired');
 
-  var diff, domain, first, threw;
-  while (first = L.peek(unrefList)) {
-    diff = now - first._idleStart;
+  var timeSinceLastActive;
+  var nextTimeoutTime;
+  var nextTimeoutDuration;
+  var minNextTimeoutTime = TIMEOUT_MAX;
+  var timersToTimeout = [];
+
+  // The actual timer fired and has not yet been rearmed,
+  // let's consider its next firing time is invalid for now.
+  // It may be set to a relevant time in the future once
+  // we scanned through the whole list of timeouts and if
+  // we find a timeout that needs to expire.
+  unrefTimer.when = -1;
 
-    if (diff < first._idleTimeout) {
-      diff = first._idleTimeout - diff;
-      unrefTimer.start(diff, 0);
-      unrefTimer.when = now + diff;
-      debug('unrefTimer rescheudling for later');
-      return;
+  // Iterate over the list of timeouts,
+  // call the onTimeout callback for those expired,
+  // and rearm the actual timer if the next timeout to expire
+  // will expire before the current actual timer.
+  var cur = unrefList._idlePrev;
+  while (cur !== unrefList) {
+    timeSinceLastActive = now - cur._idleStart;
+
+    if (timeSinceLastActive < cur._idleTimeout) {
+      // This timer hasn't expired yet, but check if its expiring time is
+      // earlier than the actual timer's expiring time
+
+      nextTimeoutDuration = cur._idleTimeout - timeSinceLastActive;
+      nextTimeoutTime = now + nextTimeoutDuration;
+      if (minNextTimeoutTime === TIMEOUT_MAX ||
+          (nextTimeoutTime < minNextTimeoutTime)) {
```

Comment:
`nextTimeoutTime` is always going to be smaller when `minNextTimeoutTime === TIMEOUT_MAX`, right?  I think you should be able to simplify the check here.